src/IO/WriteBufferFromS3TaskTracker.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include "WriteBufferFromS3.h"
+
+namespace DB
+{
+
+class WriteBufferFromS3::TaskTracker
+{
+public:
+    using Callback = std::function<void()>;
+
+    explicit TaskTracker(ThreadPoolCallbackRunner<void> scheduler_);
+    ~TaskTracker();
+
+    static ThreadPoolCallbackRunner<void> syncRunner();
+
+    bool isAsync() const;
+    void getReady();
+    void getAll();
+    void safeWaitAll();
+    void add(Callback && func);
+
+private:
+    bool is_async;
+    ThreadPoolCallbackRunner<void> scheduler;
+    std::list<std::future<void>> futures;
```

Comment:
The think I don't like about std::future is that it contains [exception, mutex and condvar](https://github.com/ClickHouse/libcxx/blob/172b2ae074f6755145b91c53a95c8540c1468239/include/future#L536-L539). Creating it for every task may be a bit overkill (but probably ok for this case).

Reply:
I share your opinion. But it is created anyway by ThreadPoolCallbackRunner<void>. Packaged task stores promise and correctly wraps all the exceptions to the future.
As I understand. All that staff is stored in the associated state which is shared between future and promise. That state lives while at least either promise or future lives. The promise is preserved as long as function waits execution and execution. Therefore I just not throw all that staff and use for good.

Reply:
Indeed ... :(